staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go

```diff
@@ -231,13 +239,26 @@ func (e *EnvSet) Extend(options ...VersionedOptions) (*EnvSet, error) {
 	return e, nil
 }
 
-func (e *EnvSet) filterAndBuildOpts(base *cel.Env, compatVer *version.Version, opts []VersionedOptions) (cel.EnvOption, error) {
+func (e *EnvSet) filterAndBuildOpts(base *cel.Env, compatVer *version.Version, honorFeatureGateEnablement bool, opts []VersionedOptions) (cel.EnvOption, error) {
 	var envOpts []cel.EnvOption
 	var progOpts []cel.ProgramOption
 	var declTypes []*apiservercel.DeclType
 
 	for _, opt := range opts {
+		var allowedByFeatureGate, allowedByVersion bool
+		if opt.FeatureEnabled != nil && honorFeatureGateEnablement {
+			// Feature-gate-enabled libraries must follow compatible default feature enablement.
+			// Enabling alpha features in their first release enables libraries the previous API server is unaware of.
+			allowedByFeatureGate = opt.FeatureEnabled()
+			if !allowedByFeatureGate {
+				continue
+			}
+		}
 		if compatVer.AtLeast(opt.IntroducedVersion) && (opt.RemovedVersion == nil || compatVer.LessThan(opt.RemovedVersion)) {
+			allowedByVersion = true
+		}
+
+		if allowedByFeatureGate || allowedByVersion {
```

Comment:
```suggestion
		allowedByFeatureGate := true
		allowedByVersion := false
		if opt.FeatureEnabled != nil && honorFeatureGateEnablement {
			// Feature-gate-enabled libraries must follow compatible default feature enablement.
			// Enabling alpha features in their first release enables libraries the previous API server is unaware of.
			allowedByFeatureGate = opt.FeatureEnabled()
		}
		if compatVer.AtLeast(opt.IntroducedVersion) && (opt.RemovedVersion == nil || compatVer.LessThan(opt.RemovedVersion)) {
			allowedByVersion = true
		}

		if allowedByFeatureGate && allowedByVersion {
```

I think this is what was intended.

Reply:
OK, this doesn't help :(  The logic is equivalent so the problem is the same.

The problem is that environment extension is dropping variable additions due to:

https://github.com/kubernetes/kubernetes/blob/403301bfdf2c7312591077827abd2e72f445a53a/test/integration/controlplane/transformation/transformation_test.go#L119

(`compatVer.AtLeast(opt.IntroducedVersion` is false).

Setting `BinaryVersion: "1.0"` would seem like the obvious fix.  @siyuanfoundation are you aware of any problems this would cause?

Reply:
@siyuanfoundation pointed out to me offline that the 0.0 binary version is to needed in this particular set of test to make sure ALL versions are covered.

Instead we can set our "CEL epoch version" to 0.0 everywhere. E.g.: https://github.com/kubernetes/kubernetes/pull/125848

It works locally.

Reply:
ah, fascinating ... so... this test *should* have been failing in master, but the init()-called env initialization snuck in libraries before the test-set binary version pushed it back to 0.0 which *should* have made *all* the libraries declaring their available version as 1.0 unavailable.

> @siyuanfoundation pointed out to me offline that the 0.0 binary version is to needed in this particular set of test to make sure ALL versions are covered.

Given this test is specifically testing kube-apiserver versions and transformation which didn't exist prior to 1.0, I don't think we care about prior to 1.0.

> Instead we can set our "CEL epoch version" to 0.0 everywhere. E.g.: https://github.com/kubernetes/kubernetes/pull/125848

I'd like to minimize the delta for existing unrelated code in this PR... I'd propose fixing the test to `BinaryVersion: "1.0"` and following up if we want to change the epoch to 1.0. I think a better future pattern would be to tie libraries to feature gates rather than artificial epochs earlier than the version when they were really introduced.

Reply:
actually, I don't understand how this test is expecting to work with BinaryVersion 0.0 OR 1.0... the APIs required to actually start a kube-apiserver don't exist at those versions

Reply:
commented on the PR that added this at https://github.com/kubernetes/kubernetes/pull/122891/files#r1663277481

Reply:
https://github.com/kubernetes/kubernetes/pull/125848 cleaned up that special-case use (thanks), rebased this after that PR merged